File: Order-Processing-Frontend/src/app/services/settings.service.ts
import {ChangeNameRequest} from "../models/settings/ChangeNameRequest";
import { environment } from 'src/environments/environment';
import {Observable} from "rxjs";
import {HttpClient} from "@angular/common/http";
import {ChangePasswordRequest} from "../models/settings/ChangePasswordRequest";
import {SignInOutService} from "./sign-in-out.service";
import {Injectable} from "@angular/core";
@Injectable({
  providedIn: 'root'
})
export class SettingsService {
  constructor(private http: HttpClient, private signInOutService: SignInOutService) {}

  changeName(request: ChangeNameRequest): Observable<string> {
    return this.http.put<string>(`${environment.baseUrl}/settings/changeName`, request, {
      params: { sessionId: this.signInOutService.getSignedInUserSessionID() },
    });
  }

  changeUsername(username: string): Observable<string> {
    return this.http.put<string>(`${environment.baseUrl}/settings/changeUsername`, username, {
      params: { sessionId: this.signInOutService.getSignedInUserSessionID() },
    });
  }

  changePassword(request: ChangePasswordRequest): Observable<string> {
    return this.http.put<string>(`${environment.baseUrl}/settings/changePassword`, request, {
      params: { sessionId: this.signInOutService.getSignedInUserSessionID() },
    });
  }

  changeEmail(email: string): Observable<string> {
    return this.http.put<string>(`${environment.baseUrl}/settings/changeEmail`, email, {
      params: { sessionId: this.signInOutService.getSignedInUserSessionID() },
    });

  }

  changeAddress(address: string): Observable<string> {
    return this.http.put<string>(`${environment.baseUrl}/settings/changeAddress`, address, {
      params: { sessionId: this.signInOutService.getSignedInUserSessionID() },
    });
  }

  changePhoneNumber(phoneNumber: string): Observable<string> {
    return this.http.put<string>(`${environment.baseUrl}/settings/changePhone`, phoneNumber, {
      params: { sessionId: this.signInOutService.getSignedInUserSessionID() },
    });
  }

  changeUserName(userName: string): Observable<string> {
    return this.http.put<string>(`${environment.baseUrl}/settings/changeUserName`, userName, {
      params: { sessionId: this.signInOutService.getSignedInUserSessionID() },
    });
  }



}

File: Order-Processing-Backend/src/main/java/csed/database/orderprocessingbackend/service/SettingsService.java
package csed.database.orderprocessingbackend.service;

import csed.database.orderprocessingbackend.model.User;
import csed.database.orderprocessingbackend.model.settings.ChangeNameRequest;
import csed.database.orderprocessingbackend.model.settings.ChangePasswordRequest;
import csed.database.orderprocessingbackend.model.settings.SettingsResponses;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class SettingsService {

    private final UserService userService;
    private final ActiveUserService activeUserService = ActiveUserService.getInstance();

    public SettingsService(UserService userService) {
        this.userService = userService;
    }


    // change name
    public SettingsResponses changeName(String sessionId, ChangeNameRequest request) {
        Long userId = activeUserService.getUserIdFromSessionId(sessionId);
        if (userId == null) {
            return SettingsResponses.INVALID_SESSION_ID;
        }
        Optional<User> user = userService.getUserById(userId);
        if(user.isPresent()) {
            userService.updateUserFirstName(request.getFirstName(), userId);
            userService.updateUserLastName(request.getLastName(), userId);
            return SettingsResponses.SUCCESSFUL;
        } else {
            return SettingsResponses.FAILED;
        }
    }


    public SettingsResponses changePassword(String sessionId, ChangePasswordRequest request) {
        Long userId = activeUserService.getUserIdFromSessionId(sessionId);
        if (userId == null) {
            return SettingsResponses.INVALID_SESSION_ID;
        }
        Optional<User> user = userService.getUserById(userId);
        if(user.isPresent()) {
            if (user.get().getPassword().equals(request.getCurrentPassword())) {
                userService.updateUserPassword(request.getNewPassword(), userId);
                return SettingsResponses.SUCCESSFUL;
            } else {
                return SettingsResponses.FAILED;
            }
        } else {
            return SettingsResponses.FAILED;
        }
    }

    public SettingsResponses changePhoneNumber(String sessionId, String phoneNumber) {
        Long userId = activeUserService.getUserIdFromSessionId(sessionId);
        if (userId == null) {
            return SettingsResponses.INVALID_SESSION_ID;
        }
        Optional<User> user = userService.getUserById(userId);
        if(user.isPresent()) {
            userService.updateUserPhoneNumber(phoneNumber, userId);
            return SettingsResponses.SUCCESSFUL;
        } else {
            return SettingsResponses.FAILED;
        }
    }

    public SettingsResponses changeEmail(String sessionId, String email) {
        Long userId = activeUserService.getUserIdFromSessionId(sessionId);
        if (userId == null) {
            return SettingsResponses.INVALID_SESSION_ID;
        }
        Optional<User> user = userService.getUserById(userId);
        if(user.isPresent()) {
            // check if email is already taken
            if (userService.getUserByEmail(email).isPresent()) {
                return SettingsResponses.EMAIL_ALREADY_TAKEN;
            }
            userService.updateUserEmail(email, userId);
            return SettingsResponses.SUCCESSFUL;
        } else {
            return SettingsResponses.FAILED;
        }
    }

    public SettingsResponses changeAddress(String sessionId, String address) {
        Long userId = activeUserService.getUserIdFromSessionId(sessionId);
        if (userId == null) {
            return SettingsResponses.INVALID_SESSION_ID;
        }
        Optional<User> user = userService.getUserById(userId);
        if(user.isPresent()) {
            userService.updateUserAddress(address, userId);
            return SettingsResponses.SUCCESSFUL;
        } else {
            return SettingsResponses.FAILED;
        }
    }

    public SettingsResponses changeUserName(String sessionId, String username) {
        Long userId = activeUserService.getUserIdFromSessionId(sessionId);
        if (userId == null) {
            return SettingsResponses.INVALID_SESSION_ID;
        }
        Optional<User> user = userService.getUserById(userId);
        if(user.isPresent()) {
            // check if username already taken
            if (userService.getUserByUsername(username).isPresent()) {
                return SettingsResponses.USERNAME_ALREADY_TAKEN;
            }
            userService.updateUsername(username, userId);
            return SettingsResponses.SUCCESSFUL;
        } else {
            return SettingsResponses.FAILED;
        }
    }
}
